// import-js/build/initializeLogging.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _winston = _interopRequireDefault(require("winston"));

var _Configuration = _interopRequireDefault(require("./Configuration"));

var _rerouteConsoleLog = _interopRequireDefault(require("./rerouteConsoleLog"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function initializeLogging(pathToLogFile) {
  // The `importjs` here is mostly a dummy file because config relies on a
  // `pathToCurrentFile`. Normally, this is the javascript file you are
  // editing.
  var level = new _Configuration.default('importjs').get('logLevel');

  _winston.default.configure({
    level: level,
    transports: [new _winston.default.transports.File({
      filename: pathToLogFile,
      json: false,
      timestamp: function timestamp() {
        return Date.now();
      },
      formatter: function formatter(_ref) {
        var timestamp = _ref.timestamp,
            level = _ref.level,
            message = _ref.message;
        var parts = [timestamp(), level.toUpperCase()];

        if (initializeLogging.parentPid) {
          // This gets set when run as a daemon
          parts.push("PID:".concat(initializeLogging.parentPid));
        }

        parts.push(message);
        return parts.join(' ');
      }
    })]
  });

  (0, _rerouteConsoleLog.default)();
}

var _default = initializeLogging;
exports.default = _default;
